### PR TITLE
proxyd: add support for limiting max block range

### PR DESCRIPTION
**Pull request by mdehoog**
**Description**

We've noticed some folks calling eth_getLogs and other RPC methods with large block ranges, which can cause our nodes to OOM. This PR introduces the ability for proxyd to block requests that have a large block range, similar to other node providers.

**Tests**

Added tests for the rewriter.

### CODE REVIEW

File: proxyd/server.go
Comment (mdehoog) [inline proxyd/server.go]: TODO: guard concurrent reads (or use atomic package)

File: proxyd/server.go
Comment (refcell) [inline proxyd/server.go]: Looks like this is missing a `continue` here unless I'm misunderstanding
```suggestion
		    continue
		}
```

File: proxyd/server.go
Comment (refcell) [inline proxyd/server.go]: It looks like we're holding this mutex across the entire function. It might make more sense to split the `updateLatestBlockNumber` function into a `fetchLatestBlockNumber() (uint64, error)` and then a `updateLatestBlockNumber(uint64)` that contains the locking logic. Something like:

```golang
func (s *Server) pollLatestBlockNumber() {
	ticker := time.NewTicker(2 * time.Second)
	for {
		select {
		case <-ticker.C:
			bn, err := s.fetchLatestBlockNumber()
            if err != nil {
                s.updateLatestBlockNumber(bn)
            }
		}
	}
}

func (s *Server) updateLatestBlockNumber(bn uint64) {
    s.srvMu.Lock()
    defer s.srvMu.Unlock()
	s.latestBlockNumber.Store(bn)
}

func (s *Server) fetchLatestBlockNumber() (uint64, error) {
	req := json.RawMessage("{\"id\":0,\"jsonrpc\":\"2.0\",\"method\":\"eth_blockNumber\"}")
	res, _, err := s.handleBatchRPC(context.Background(), []json.RawMessage{req}, func(method string) bool { return false }, false)
	if err != nil {
		log.Error("error requesting latest block number", "err", err)
		return uint64(0), err
	}
	bns := res[0].Result.(string)
	bn, err := hexutil.DecodeUint64(bns)
	if err != nil {
		log.Error("error decoding hex block number", "err", err)
		return uint64(0), err
	}
	return bn, nil
}
```

Comment (mdehoog) [reply proxyd/server.go]: The purpose of using the mutex is to guard against calling `Shutdown` during a call to `updateLatestBlockNumber()`, which would cause the backend groups to shutdown and therefore the `handleBatchRPC` could panic. It also guards the check of the `shutdown` boolean so the goroutine can exit. But if we don't care about this during shutdown, we can remove

File: proxyd/server.go
Comment (mdehoog) [reply proxyd/server.go]: good catch! thanks

File: proxyd/server.go
Comment (refcell) [reply proxyd/server.go]: SGTM, will dismiss this, still works the current way so no concern here :)

File: proxyd/latest_block_poller.go
Comment (refcell) [inline proxyd/latest_block_poller.go]: Note, a `for range` loop over a receiving channel will _not_ execute until the first interval is passed. In other words, it will first wait 2 seconds and then perform the first poll.

To instead have this execute immediately, we could do something like the following, which will run the loop immediately and wait on receive for each subsequent iteration.
```suggestion
    for ; true; <-ticker.C {
		p.mutex.Lock()
		if p.shutdown {
			ticker.Stop()
			p.mutex.Unlock()
			return
		}
		p.poll()
		p.mutex.Unlock()
	}
```

File: proxyd/latest_block_poller.go
Comment (refcell) [inline proxyd/latest_block_poller.go]: The comment below addressing the `start` method would allow you to remove this additional `p.poll()` call.

File: proxyd/latest_block_poller.go
Comment (refcell) [inline proxyd/latest_block_poller.go]: Might make sense to have this `2 * time.Second` duration be configurable as a `LatestBlockPoller` field and argument to the `NewLatestBlockPoller` constructor.

**Review by refcell — COMMENTED** [review]
Looking great!

File: proxyd/latest_block_poller.go
Comment (mdehoog) [reply proxyd/latest_block_poller.go]: It's a minor thing, but i made the `poll` call in the original goroutine (and allowed the interval to pass in the separate one) so that the latest block has a chance to be populated before returning. But given the handling around 0 blocks, it probably doesn't matter too much. Let me know if you think we should switch to your suggestion.

File: proxyd/latest_block_poller.go
Comment (mdehoog) [reply proxyd/latest_block_poller.go]: done!

File: proxyd/latest_block_poller.go
Comment (trianglesphere) [reply proxyd/latest_block_poller.go]: I lean to this being fine. The majority of the time proxyd is running, it will have this value populated, & if it doesn't it's not the end of the world.

**Review by trianglesphere — COMMENTED** [review]
The poller & `IsAllowedBlockRange` look good to me. Because I'm not super familiar with this code, I'd like eyes from someone else to confirm that we're running `IsAllowedBlockRange` in all of the places that we need to be running it.

File: proxyd/latest_block_poller.go
Comment (mdehoog) [reply proxyd/latest_block_poller.go]: sgtm, will update

File: proxyd/latest_block_poller.go
Comment (mdehoog) [reply proxyd/latest_block_poller.go]: done

**Comment by felipe-op** [comment]
Please take a look at the consensus poller: https://github.com/ethereum-optimism/optimism/blob/develop/proxyd/consensus_poller.go
It is already updating the latest block in a configurable interval. This change seems to duplicate the same behaviour.

You can enable consensus awareness in the backend group configs:
https://github.com/ethereum-optimism/optimism/blob/edafea3e21251e4efbc3075471e6e4a0e088cdd7/proxyd/example.config.toml#L94

Once enabled, block tags such as `latest` will be rewritten by their correspondent block number: https://github.com/ethereum-optimism/optimism/blob/edafea3e21251e4efbc3075471e6e4a0e088cdd7/proxyd/rewriter.go#L117

And you then can just validate the numerical block range interval without worrying with tracking `latest` or converting the block tags to block numbers.

**Review by felipe-op — CHANGES_REQUESTED** [review]
This change should integrate with consensus awareness

File: proxyd/backend.go
Comment (mdehoog) [inline proxyd/backend.go:228]: Allows a backend to opt-out of all the other consensus health checks (sync, peers, healthy, banned, etc). This is important for us as we only have a single backend and we never want it to not be a candidate.

File: proxyd/rewriter.go
Comment (mdehoog) [inline proxyd/rewriter.go:15]: should the `maxBlockRange` be a field in the `RewriteContext` struct, or passed in as a parameter to the `RewriteTags` / `RewriteRequest` / `rewriteRange` methods?

this way seems cleaner to me

File: proxyd/rewriter.go
Comment (mdehoog) [inline proxyd/rewriter.go:133]: this avoids the situation where somebody uses `fromBlock = 0` but leaves `toBlock` blank, which might be ignored by the proxyd range check, but `toBlock` defaults to `latest` in geth

**Comment by mdehoog** [comment]
@felipe-op great suggestion, reimplemented 👍

File: proxyd/backend.go
Comment (mslipper) [inline proxyd/backend.go]: [Nit] Should be an invalid params error - can use `ErrInvalidParams()` to generate this.

File: proxyd/rewriter_test.go
Comment (mslipper) [inline proxyd/rewriter_test.go:151]: Recommend some additional tests

- Range from `earliest` to `latest`
- No `toBlock` defined
- No `fromBlock` defined

**Review by mslipper — APPROVED** [review]
LGTM once testing/nits are addressed.

File: proxyd/rewriter_test.go
Comment (mdehoog) [reply proxyd/rewriter_test.go:151]: good suggestion on `earliest`, picked up a bug where `rewriteTags` doesn't rewrite `earliest` or `pending`

**Review by felipe-op — APPROVED** [review]

**Comment by felipe-op** [comment]
Please squash before merging

**Comment by mdehoog** [comment]
thanks... squashed 👍

**Comment by mdehoog** [comment]
(pushed a no-op commit to re-run tests, seems flaky)

**Comment by mslipper** [comment]
Merged in upstream - that should fix the tests.

**Pull request merged**
